File: rb/test/entity_template_entity_test.rb
# EntityTemplate entity test

require "minitest/autorun"
require "json"
require_relative "../Shortcut_sdk"
require_relative "runner"

class EntityTemplateEntityTest < Minitest::Test
  def test_create_instance
    testsdk = ShortcutSDK.test(nil, nil)
    ent = testsdk.EntityTemplate(nil)
    assert !ent.nil?
  end

  # Feature #4: the entity stream(action, ...) method runs the op pipeline and
  # returns an Enumerator over result items. With the streaming feature active
  # it yields the feature's incremental output; otherwise it falls back to the
  # materialised list so stream always yields.
  def test_stream
    seed = {
      "entity" => {
        "entity_template" => {
          "s1" => { "id" => "s1" },
          "s2" => { "id" => "s2" },
          "s3" => { "id" => "s3" },
        },
      },
    }

    # Fallback: streaming inactive -> yields the materialised list items.
    base = ShortcutSDK.test(seed, nil)
    seen = base.EntityTemplate(nil).stream("list", nil, nil).to_a
    assert_equal 3, seen.length

    # Inbound: streaming active -> yields each item from the feature.
    cfg = ShortcutConfig.shared_config
    if cfg["feature"].is_a?(Hash) && cfg["feature"].key?("streaming")
      sdk = ShortcutSDK.test(seed, { "feature" => { "streaming" => { "active" => true } } })
      got = []
      sdk.EntityTemplate(nil).stream("list", nil, nil).each do |item|
        if item.is_a?(Array)
          got.concat(item)
        else
          got << item
        end
      end
      assert_equal 3, got.length
    end
  end

  def test_basic_flow
    setup = entity_template_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create", "list", "update", "load", "remove"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "entity_template." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set SHORTCUT_TEST_ENTITY_TEMPLATE_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    entity_template_ref01_ent = client.EntityTemplate(nil)
    entity_template_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.entity_template"), "entity_template_ref01"))

    entity_template_ref01_data_result = entity_template_ref01_ent.create(entity_template_ref01_data, nil)
    entity_template_ref01_data = Helpers.to_map(entity_template_ref01_data_result.respond_to?(:data_get) ? entity_template_ref01_data_result.data_get : entity_template_ref01_data_result)
    assert !entity_template_ref01_data.nil?
    assert !entity_template_ref01_data["id"].nil?

    # LIST
    entity_template_ref01_match = {}

    entity_template_ref01_list_result = entity_template_ref01_ent.list(entity_template_ref01_match, nil)
    assert entity_template_ref01_list_result.is_a?(Array)

    found_item = Vs.select(
      Runner.entity_list_to_data(entity_template_ref01_list_result),
      { "id" => entity_template_ref01_data["id"] })
    assert !Vs.isempty(found_item)

    # UPDATE
    entity_template_ref01_data_up0_up = {
      "id" => entity_template_ref01_data["id"],
    }

    entity_template_ref01_markdef_up0_name = "author_id"
    entity_template_ref01_markdef_up0_value = "Mark01-entity_template_ref01_#{setup[:now]}"
    entity_template_ref01_data_up0_up[entity_template_ref01_markdef_up0_name] = entity_template_ref01_markdef_up0_value

    entity_template_ref01_resdata_up0_result = entity_template_ref01_ent.update(entity_template_ref01_data_up0_up, nil)
    entity_template_ref01_resdata_up0 = Helpers.to_map(entity_template_ref01_resdata_up0_result.respond_to?(:data_get) ? entity_template_ref01_resdata_up0_result.data_get : entity_template_ref01_resdata_up0_result)
    assert !entity_template_ref01_resdata_up0.nil?
    assert_equal entity_template_ref01_resdata_up0["id"], entity_template_ref01_data_up0_up["id"]
    assert_equal entity_template_ref01_resdata_up0[entity_template_ref01_markdef_up0_name], entity_template_ref01_markdef_up0_value

    # LOAD
    entity_template_ref01_match_dt0 = {
      "id" => entity_template_ref01_data["id"],
    }
    entity_template_ref01_data_dt0_loaded = entity_template_ref01_ent.load(entity_template_ref01_match_dt0, nil)
    entity_template_ref01_data_dt0_load_result = Helpers.to_map(entity_template_ref01_data_dt0_loaded.respond_to?(:data_get) ? entity_template_ref01_data_dt0_loaded.data_get : entity_template_ref01_data_dt0_loaded)
    assert !entity_template_ref01_data_dt0_load_result.nil?
    assert_equal entity_template_ref01_data_dt0_load_result["id"], entity_template_ref01_data["id"]

    # REMOVE
    entity_template_ref01_match_rm0 = {
      "id" => entity_template_ref01_data["id"],
    }
    entity_template_ref01_ent.remove(entity_template_ref01_match_rm0, nil)

    # LIST
    entity_template_ref01_match_rt0 = {}

    entity_template_ref01_list_rt0_result = entity_template_ref01_ent.list(entity_template_ref01_match_rt0, nil)
    assert entity_template_ref01_list_rt0_result.is_a?(Array)

    not_found_item = Vs.select(
      Runner.entity_list_to_data(entity_template_ref01_list_rt0_result),
      { "id" => entity_template_ref01_data["id"] })
    assert Vs.isempty(not_found_item)

  end
end

def entity_template_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "entity_template", "EntityTemplateTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = ShortcutSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["entity_template01", "entity_template02", "entity_template03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["SHORTCUT_TEST_ENTITY_TEMPLATE_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "SHORTCUT_TEST_ENTITY_TEMPLATE_ENTID" => idmap,
    "SHORTCUT_TEST_LIVE" => "FALSE",
    "SHORTCUT_TEST_EXPLAIN" => "FALSE",
    "SHORTCUT_APIKEY" => "",
  })

  idmap_resolved = Helpers.to_map(
    env["SHORTCUT_TEST_ENTITY_TEMPLATE_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["SHORTCUT_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      # FIRST, so the generated fields below win: sdk-test-control.json's
      # test.client.options adds to the live client, it does not redirect it.
      Runner.live_client_options,
      {
        "apikey" => env["SHORTCUT_APIKEY"],
      },
      extra || {},
    ])
    client = ShortcutSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["SHORTCUT_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["SHORTCUT_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: go/test/entity_template_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/shortcut-sdk/go"
	"github.com/voxgig-sdk/shortcut-sdk/go/core"

	vs "github.com/voxgig-sdk/shortcut-sdk/go/utility/struct"
)

func TestEntityTemplateEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.EntityTemplate(nil)
		if ent == nil {
			t.Fatal("expected non-nil EntityTemplateEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"entity_template": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.EntityTemplate(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.EntityTemplate(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := entity_templateBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "entity_template." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set SHORTCUT_TEST_ENTITY_TEMPLATE_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		entityTemplateRef01Ent := client.EntityTemplate(nil)
		entityTemplateRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "entity_template"}), "entity_template_ref01"))

		entityTemplateRef01DataResult, err := entityTemplateRef01Ent.Create(entityTemplateRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		entityTemplateRef01Data = core.ToMapAny(entityData(entityTemplateRef01DataResult))
		if entityTemplateRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if entityTemplateRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		entityTemplateRef01Match := map[string]any{}

		entityTemplateRef01ListResult, err := entityTemplateRef01Ent.List(entityTemplateRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		entityTemplateRef01List, entityTemplateRef01ListOk := entityTemplateRef01ListResult.([]any)
		if !entityTemplateRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", entityTemplateRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(entityTemplateRef01List), map[string]any{"id": entityTemplateRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		entityTemplateRef01DataUp0Up := map[string]any{
			"id": entityTemplateRef01Data["id"],
		}

		entityTemplateRef01MarkdefUp0Name := "author_id"
		entityTemplateRef01MarkdefUp0Value := fmt.Sprintf("Mark01-entity_template_ref01_%d", setup.now)
		entityTemplateRef01DataUp0Up[entityTemplateRef01MarkdefUp0Name] = entityTemplateRef01MarkdefUp0Value

		entityTemplateRef01ResdataUp0Result, err := entityTemplateRef01Ent.Update(entityTemplateRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		entityTemplateRef01ResdataUp0 := core.ToMapAny(entityData(entityTemplateRef01ResdataUp0Result))
		if entityTemplateRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if entityTemplateRef01ResdataUp0["id"] != entityTemplateRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if entityTemplateRef01ResdataUp0[entityTemplateRef01MarkdefUp0Name] != entityTemplateRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", entityTemplateRef01MarkdefUp0Name, entityTemplateRef01ResdataUp0[entityTemplateRef01MarkdefUp0Name])
		}

		// LOAD
		entityTemplateRef01MatchDt0 := map[string]any{
			"id": entityTemplateRef01Data["id"],
		}
		entityTemplateRef01DataDt0Loaded, err := entityTemplateRef01Ent.Load(entityTemplateRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		entityTemplateRef01DataDt0LoadResult := core.ToMapAny(entityData(entityTemplateRef01DataDt0Loaded))
		if entityTemplateRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if entityTemplateRef01DataDt0LoadResult["id"] != entityTemplateRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		entityTemplateRef01MatchRm0 := map[string]any{
			"id": entityTemplateRef01Data["id"],
		}
		_, err = entityTemplateRef01Ent.Remove(entityTemplateRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		entityTemplateRef01MatchRt0 := map[string]any{}

		entityTemplateRef01ListRt0Result, err := entityTemplateRef01Ent.List(entityTemplateRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		entityTemplateRef01ListRt0, entityTemplateRef01ListRt0Ok := entityTemplateRef01ListRt0Result.([]any)
		if !entityTemplateRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", entityTemplateRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(entityTemplateRef01ListRt0), map[string]any{"id": entityTemplateRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func entity_templateBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "entity_template", "EntityTemplateTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read entity_template test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse entity_template test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"entity_template01", "entity_template02", "entity_template03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("SHORTCUT_TEST_ENTITY_TEMPLATE_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"SHORTCUT_TEST_ENTITY_TEMPLATE_ENTID": idmap,
		"SHORTCUT_TEST_LIVE":      "FALSE",
		"SHORTCUT_TEST_EXPLAIN":   "FALSE",
		"SHORTCUT_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["SHORTCUT_TEST_ENTITY_TEMPLATE_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["SHORTCUT_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["SHORTCUT_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewShortcutSDK(core.ToMapAny(mergedOpts))
	}

	live := env["SHORTCUT_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["SHORTCUT_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
